src/proxy-restore.c

```diff
@@ -11,24 +10,44 @@
 // call `vec_clone_referenced()`, which won't attempt to clone if we know we
 // own the object. See #1151.
 r_obj* vec_restore(r_obj* x, r_obj* to, const enum vctrs_owned owned) {
-  switch (class_type(to)) {
+  return vec_restore_4(x, to, owned, VCTRS_RECURSE_false);
+}
+r_obj* vec_restore_recurse(r_obj* x, r_obj* to, const enum vctrs_owned owned) {
+  return vec_restore_4(x, to, owned, VCTRS_RECURSE_true);
+}
+
+r_obj* ffi_vec_restore(r_obj* x, r_obj* to) {
+  return vec_restore_4(x, to, vec_owned(x), VCTRS_RECURSE_false);
+}
+r_obj* ffi_vec_restore_recurse(r_obj* x, r_obj* to) {
+  return vec_restore_4(x, to, vec_owned(x), VCTRS_RECURSE_true);
+}
+
+static
+r_obj* vec_restore_4(r_obj* x,
+                     r_obj* to,
+                     const enum vctrs_owned owned,
+                     bool recurse) {
+  enum vctrs_class_type to_type = class_type(to);
+
+  if (recurse && !class_type_is_data_frame(to_type) && is_data_frame(x)) {
+    return vec_df_restore(x, to, owned, recurse);
+  }
```

Comment:
This is the only section I am still a little uncomfortable with. I wrote a lot here just to have it written down somewhere, and I _think_ by the end I convinced myself that the pros of your approach outweigh any cons.

---

I _feel_ that I as the class creator should be in complete control of the proxy/restore path. Something about _automatic_ recursive proxy / restore feels wrong here, whereas it makes sense for `vec_proxy_equal()` and those variants (probably because those are less performance sensitive, and there is no restoration needed).

I think a big issue for me is that it feels wrong that we have to `vec_proxy(to)` to be able to do the recursive restoration. I am a little worried about the performance implications of that, because proxy/restore methods aren't the fastest things in the world. Although, hopefully the fact that we'd only call `vec_restore_recurse()` once at the end of `vec_c/rbind/cbind()` will help with performance. 

I'm going to list a few cases where I could see the recursion coming into play, and I'm going to talk about how they could work with the `recurse = TRUE` argument being passed through the `...`.

- Assume `x` is a bare rcrd
  - `vec_proxy.vctrs_rcrd(x, ...)` could automatically recursively proxy the fields of `x` if `recurse = TRUE` is passed through `...`. This is the best default we can supply without knowing more about the specific rcrd type, and means that most rcrd subclasses don't need any proxy methods.
  - `vec_restore.vctrs_rcrd(x, to, ...)` could automatically recursively restore the fields of `x` to the fields in `to` if `recurse = TRUE`. Since we know `to` is a rcrd and we know `x` is a data frame with the same number of columns as in `to`, we don't need to `vec_proxy(to)` to be able to restore `x` to the type of `to` recursively. We just call `vec_restore_recursive()` on each column/field. If a user supplies a rcrd subclass proxy method that changes the number of columns, then it is up to them to also supply a restore method, but I think this is unlikely.

- Assume `x` is an iv
  - I don't know what the `start` and `end` field types are, so they will need to be recursively proxied/restored. Luckily, the rcrd method handles this for me so I don't need to care about that.
  - There would not be `vec_proxy.iv()` or `vec_restore.iv()` methods.

- Assume `x` is a clock type like year_month_day
  - I know _exactly_ what each field is. They are all integer vectors that don't need recursive proxy/restore.
  - I can choose to not write proxy/restore methods and everything will work fine.
  - Or I can choose to write "performance" methods for proxy/restore and choose to ignore the `recurse = TRUE` option passed through `...` since my fields don't need to be proxied or restored.

- Assume `x` isn't a rcrd or data frame, but it does have a data frame proxy.
  - I think most of the time this will be fine, the data frame columns are probably atomic types.
  - I'd argue that if the data frame proxy could have unknown column types, then it is the class creator's job to check for `recurse = TRUE` being passed through the `...` and then call `vec_proxy(col, ...)` on each column if we are doing recursive proxying. (I think this is a red flag for this approach, because the method writer will have to look at the internal `recurse` option to know if they should proxy each column or not).
  - Same argument with `vec_restore()`, but it is likely that the class writer will have enough knowledge about `to` that they won't have to call `vec_proxy(to)` to be able to restore to the type of `to`.

---

Even after writing out these 4 cases, I'm not completely sure what we should do.

Pros of both:
- I like the idea that most rcrd subclass writers shouldnt have to worry about this. Both approaches give us that.

Pros of `recurse = TRUE` through `...`:
- I like the idea of "performance" methods that can decide to ignore `recurse` if the fields don't need it. Automatic recursive proxy/restore means we can't short circuit the recursion for performance.
- I like the idea of not needing to call `vec_proxy(to)` during restoration. Explicit handling of `recurse` gives class writers a chance to use their knowledge of `to` to not need this. This might not matter much since `vec_restore_recurse()` is only called 1 time at the end of the concatenation methods now.

Pros of automatic proxy/restore:
- I like the idea that if a non-rcrd has a data frame proxy, then its proxy/restore method is still applied recursively automatically for the creator. The current implementation gives us that.
  - The `recurse = TRUE` through `...` method is actually very confusing here. The method writer in theory shouldn't need to know about this argument at all, but in fact they _do_ need to know about its value to know whether or not they should call `vec_proxy(col, ...)` on each column to do the 1st layer of the recursion.
- I like that we don't need to introduce a new internal option through `...`. That will probably make it a little easier to "invert the loops" in the future, where we theoretically might remove that option anyways.

I think avoiding adding `recurse = TRUE` as an "internal" option that method writers do actually have to look at is probably the best approach (both for now and thinking towards the "invert the loop" idea).

Reply:
Thanks for sharing these. Quick summary of my thoughts:

- I like that class authors don't need to think about recursivity.
- I'm not too worried about the performance of proxying `to` at the end because it's a scalar operation, not a vectorised one.
- I like that there is no behaviour change and we are free to use another approach when we work on this problem again for the 1.0 release. This is especially important to me because we don't know for sure what the final algorithm will look like.

Reply:
Agreed that those 3 points are important enough that this approach feels right until we can revisit for 1.0, thanks!